test network package

phoracek opened the pull request:
Finish test coverage for pkg/network. This patch also fixes a bug in multus
validation where we tried to compared bool pointers.

phoracek commented:
@booxter @SchSeba @qinqon could you please take a look?

booxter commented on pkg/network/sriov_test.go:
```diff
@@ -0,0 +1,156 @@
+			})
+		})
+
+		Context("when openshfit network operator is running", func() {
```
openshift

booxter commented on pkg/network/sriov_test.go:
```diff
@@ -0,0 +1,156 @@
+					It("should fail with an error message", func() {
+						errorList := _validateSriov(sriovRequested, openshiftNetworkOperatorRunning, openshiftSriovEnabled)
+						Expect(len(errorList)).To(Equal(1), "validation failed due to an unexpected error: %v", errorList)
+						Expect(errorList[0].Error()).To(Equal("SR-IOV has been requested, but it's not compatible with OpenShift Cluster Network Operator SR-IOV support"))
```
(no action required) personally not a fan of matching against exact error messages but I guess it's subjective

phoracek replied:
Is there something like "error type" in Go? Like checking Exception class in Python?

booxter replied:
I don't think it can be directly mimicked. I don't know if there is a good solution. Is the exact message important enough to check in the first place?

phoracek replied:
I'd say it does no harm. I'd rather change the message at two places than have ambiguous test.

booxter commented on pkg/network/sriov_test.go:
```diff
@@ -0,0 +1,156 @@
+			})
+		})
+
+		Context("when openshfit network operator is not running", func() {
```
openshift

booxter commented on pkg/network/sriov_test.go:
```diff
@@ -0,0 +1,156 @@
+
+		Context("when openshfit network operator is not running", func() {
+			openshiftNetworkOperatorRunning := false
+			openshiftNetworkOperatorDisableMultiNetwork := false
```
the name of this variable differs from what is used above, and it doesn't seem like matching the meaning of the helper function parameter (?): the function parameter means "whether openshift sriov network is enabled" while the variable name suggests the meaning is "whether multi-network support (?) is disabled"?

phoracek replied:
I see, I instantiated it just so there is a value, it doesn't matter whether it's true or false, since operator is disabled. Adding a comment.

openshift's Network type has attribute DisableMultiNetwork. I don't like negative parameters, so I'm trying to hide it from upper level (eyes of a developer running these tests).

booxter replied:
it seems to me that the variable name doesn't belong here and was copy pasted from another test module (for multus).

phoracek replied:
Should be better now

booxter commented on pkg/network/nmstate_test.go:
```diff
@@ -0,0 +1,49 @@
+			})
+		})
+
+		Context("when the previous and new configuration matches", func() {
```
matches -> match

booxter commented on pkg/network/network_test.go:
```diff
@@ -0,0 +1,128 @@
+		})
+	})
+
+	Describe("FillDefaults", func() {
```
I probably see why you don't have a negative case here (because the only way to receive an error from the function is if rand.Read() fails, which is hard to simulate); but in case you care to cover all branches (which seems like you effectively do with this PR), you could introduce a test mock / helper function for rand.Read that you could replace for test scenarios to trigger an error.

phoracek replied:
I will just add a TODO, not that I am lazy, but I want to leave a low-hanging fruit for others (or future me)

booxter commented on pkg/network/network_test.go:
```diff
@@ -0,0 +1,128 @@
+			openshiftNetworkConf := &osv1.Network{}
+			clusterInfo := &ClusterInfo{SCCAvailable: true, OpenShift4: false}
+
+			It("should successfully render a set of objects", func() {
```
perhaps you could add a negative case here, this seems like an important case to cover.

phoracek replied:
Done for invalid manifestsDir

booxter commented on pkg/network/multus.go:
```diff
@@ -21,7 +21,7 @@ func validateMultus(conf *opv1alpha1.NetworkAddonsConfigSpec, openshiftNetworkCo
 	}
 
 	if openshiftNetworkConfig != nil {
-		if openshiftNetworkConfig.Spec.DisableMultiNetwork == newTrue() {
```
perhaps you should kill newTrue now then? it's not used anywhere else.

booxter commented on pkg/network/multus_test.go:
```diff
@@ -0,0 +1,147 @@
+			if openshiftNetworkOperatorRunning {
+				openshiftNetworkConfig = &osv1.Network{}
+				if openshiftNetworkOperatorDisableMultiNetwork {
+					newTrue := func() *bool {
```
I probably just don't know Golang well but what's this construction for? Can't we just have it as a simple struct field?

phoracek replied:
You cannot assign &true. You have to do this extra step to get pointer to bool. Don't know why, don't know about any other nicer way.

booxter commented on pkg/network/multus_test.go:
```diff
@@ -0,0 +1,147 @@
+			})
+		})
+
+		Context("when openshfit network operator is running", func() {
```
openshift

booxter commented on pkg/network/multus_test.go:
```diff
@@ -0,0 +1,147 @@
+			})
+		})
+
+		Context("when openshfit network operator is not running", func() {
```
openshift

booxter commented on pkg/network/multus_test.go:
```diff
@@ -0,0 +1,147 @@
+			})
+		})
+
+		Context("when the previous and new configuration matches", func() {
```
matches -> match

booxter commented on pkg/network/linux-bridge_test.go:
```diff
@@ -0,0 +1,49 @@
+			})
+		})
+
+		Context("when the previous and new configuration matches", func() {
```
matches -> match

booxter commented on pkg/network/image-pull-policy_test.go:
```diff
@@ -0,0 +1,108 @@
+			})
+		})
+
+		Context("when configration uses a valid policy type", func() {
```
configuration

booxter commented on pkg/network/image-pull-policy_test.go:
```diff
@@ -0,0 +1,108 @@
+				new := &opv1alpha1.NetworkAddonsConfigSpec{}
+				prev := &opv1alpha1.NetworkAddonsConfigSpec{}
+
+				It("should successfullt pass", func() {
```
successfully

booxter commented on pkg/network/image-pull-policy_test.go:
```diff
@@ -0,0 +1,108 @@
+				})
+
+				It("and fill in the previously defined policy", func() {
+					Expect(new.ImagePullPolicy).To(Equal(v1.PullAlways))
```
while we know that default policy is not Always, it's not strictly speaking a guarantee. And if for some reason it would be made into Always then this test case would not be distinguishable from behavior of e.g. filling it the result with default (now Always) value.

phoracek replied:
Probably my bad specifying first the new config and then old. We set PullAlways in `prev`, we don't specify any policy in `new`. There is no defaulting. Or maybe I don't understand your comment.

booxter commented on pkg/network/image-pull-policy_test.go:
```diff
@@ -0,0 +1,108 @@
+			})
+		})
+
+		Context("when the previous and new configuration matches", func() {
```
matches -> match

phoracek commented:
@booxter @qinqon addressed your comments. Resolved most.

@qinqon regarding tables - I don't like that it doesn't allow you to include `It` to it (right? or is it Context). Anyway, we can do refactoring of that in a different PR, I don't really want to touch it now if you don't mind.

Guys, sorry about being lazy ignoring some of your comments. Although I appreciate them, I'd like to get over this PR. But if you feel like fighting for those, please do!

qinqon commented:
LGTM awesome stuff
